### PR TITLE
Fix character entity handling (#450)

### LINKED ISSUE #450: bleach.clean method returns unexpected value for string with unknown entities
For string strings like "&adp;" and "&*!;" actual output is "&ad;;" and "&*;;".
But I expect that such values should remain original value ("&adp;" and "&*!;" corresponding).

### PR DESCRIPTION
**Pull request by willkg**
This fixes a bug where things in the form of a character entity (starts with &, some letters, and ends with ;) that aren't character entities get mangled. Now they are correctly handled.

This also fixes a bug where things in the form of a character entity that aren't character entities should have had the & escaped, but that wasn't happening.

After this, Bleach does the following:

* valid character entities that start with a & and end with a ; are left alone
* things that start with a & and end with a ; and aren't valid character entities are treated as text and the & is escaped
* things that start with a & but don't end with a ; are treated as text and the & is escaped

This is easy to reason about and idempotent and doesn't suffer from ambiguous edge cases where Bleach can't know what the user intended.

I also cleaned up the comments in the tests so it's clearer what we intend Bleach to be doing.

This fixes #450.

### CODE REVIEW

File: bleach/html5lib_shim.py [hunk 1]
```diff
@@ -557,13 +557,19 @@ def match_entity(stream):
...
-        if not ENTITIES_TRIE.has_keys_with_prefix(possible_entity):
-            break
+        print(1, stream, c, possible_entity)
         possible_entity += c
+        if not ENTITIES_TRIE.has_keys_with_prefix(possible_entity):
+            # If it's not a prefix, then it's not an entity and we're
+            # out
+            return None
 
+    print(2, stream, possible_entity)
```
Comment (willkg) [inline bleach/html5lib_shim.py:567]: Oops--I need to remove the print functions.

File: bleach/html5lib_shim.py [hunk 2]
```diff
@@ -557,9 +557,11 @@ def match_entity(stream):
     # Handle character entities
     while stream and stream[0] not in end_characters:
         c = stream.pop(0)
-        if not ENTITIES_TRIE.has_keys_with_prefix(possible_entity):
-            break
         possible_entity += c
+        if not ENTITIES_TRIE.has_keys_with_prefix(possible_entity):
+            # If it's not a prefix, then it's not an entity and we're
+            # out
+            return None
```
Comment (willkg) [inline bleach/html5lib_shim.py:564]: This block mirrors the structure of the one that handles numeric entities, but it was awkwardly wrong and had been since I wrote it in 2017:

https://github.com/mozilla/bleach/commit/d8f7d1d651a6b5a4c8e5a5932ea4fd61769a60d4

:(

**Review by g-k — APPROVED** [review]
r+

> This is easy to reason about and idempotent and doesn't suffer from ambiguous edge cases where Bleach can't know what the user intended.

Excellent!

Thanks for updating the test comments and writing down the expected behavior. In the future, I'd love to add property-based tests for idempotency and the three behaviors you listed.

**Comment by willkg** [comment]
Oh, that's a good idea. I'll add some more idempotent tests now.

**Comment by willkg** [comment]
@g-k ^^^ Does that cover the idempotent tests you were thinking of? Are there others we should add?

**Comment by g-k** [comment]
> Does that cover the idempotent tests you were thinking of? Are there others we should add?

Oh nice! I was thinking of it as a TODO / good predicates to check for #552

**Review by g-k — APPROVED** [review]

**Comment by willkg** [comment]
That makes sense. I haven't quite gotten around to looking at the fuzz stuff, yet. I'll toss that in my list of things to do, but probably after the 5.0.0 release.

Thank you!
